packages/mendel-pipeline/src/cache/index.js

```diff
@@ -39,21 +48,86 @@ class MendelCache extends EventEmitter {
     }
 
     addEntry(id) {
-        if (!this._store.has(id)) {
-            const entry = new Entry(id);
-            entry.variation = this.getVariation(id);
-            entry.normalizedId = this.getNormalizedId(id);
+        if (this._store.has(id)) return;
 
-            if (!this._normalizedIdToEntryIds.has(entry.normalizedId)) {
-                this._normalizedIdToEntryIds.set(entry.normalizedId, []);
-            }
-            this._normalizedIdToEntryIds.get(entry.normalizedId).push(entry.id);
+        this.handleAsPackageJSON(id);
+
+        const entry = new Entry(id);
+
+        // normalize based on variation and environment
+        entry.variation = this.getVariation(id);
+        entry.normalizedId = this.getNormalizedId(id);
+        entry.runtime = this.getRuntime(id);
+
+        // fast lookup cache per normalized id
+        if (!this._normalizedIdToEntryIds.has(entry.normalizedId)) {
+            this._normalizedIdToEntryIds.set(entry.normalizedId, []);
+        }
+        this._normalizedIdToEntryIds.get(entry.normalizedId).push(entry.id);
+
+        // finally
+        this._store.set(id, entry);
+        this.emit('entryAdded', id);
+    }
 
-            this._store.set(id, entry);
-            this.emit('entryAdded', id);
+    handleAsPackageJSON(id) {
+        const parts = path.parse(id);
+        const packageNormId = this.getNormalizedId(id);
+        const pkgPath = path.join(this.projectRoot, id);
+
+        if (parts.base !== 'package.json') return;
+        if (this._normalizedIdToEntryIds.has(packageNormId)) {
+            /*
+              This should only happen in watch mode
+
+              TODO: It is possible to not throw here.
+
+              Mutating entries is not safe, since other async pieces of the
+              pipeline might be relying on this files already.
+
+              Here are some hypothesis on how to avoid throwing:
+              1. Restart the whole process automatically when this happens
+              2. Traverse all decendants from all variations and files from
+                 this normalizedId and remove them from the pipeline and
+                 traverse all the tree again with the new package.json
+            */
+            throw new Error([
+                `can't process ${id} after the`,
+                'following files are in the system:',
+                '\n',
+                this._normalizedIdToEntryIds.get(packageNormId).join('\n'),
+                '\n',
+            ].join(' '));
         }
+
+        delete require.cache[require.resolve(pkgPath)];
+        const pkg = require(pkgPath);
+
+        ['browser', 'main'].forEach(runtime => {
+            if (!pkg[runtime]) return;
+            const targetNormId = this.getNormalizedId(
+                './' + path.join(parts.dir, pkg[runtime])
+            );
+            this._packageMap.set(targetNormId, {
```

Comment:
Can you add a small check? If `this._packageMap.has(targetNormId)`, throw saying multiple `package.json`s are pointing at the same file? Thanks!